cmd/ltkeycloak/main.go

```diff
@@ -0,0 +1,468 @@
+						Username:      &user.Username,
+						Email:         &user.Email,
+						Enabled:       gocloak.BoolP(true),
+						EmailVerified: &user.EmailVerified,
+						Credentials: &[]gocloak.CredentialRepresentation{
+							{
+								Temporary: gocloak.BoolP(false),
+								Type:      gocloak.StringP("password"),
+								Value:     gocloak.StringP(userPassword),
+							},
+						},
+					})
+					if err != nil {
+						// Ignore already existing users
+						if apiErr, ok := err.(*gocloak.APIError); ok && apiErr.Code == 409 {
+							continue
+						}
+
+						slog.Error("failed to create user in keycloak", slog.String("err", err.Error()))
+						continue
+					}
+
+					user.AuthData = model.NewString(kcUserID)
+					user.AuthService = model.UserAuthServiceSaml
+					user.Password = ""
+
+					_, _, err = mmClient.UpdateUser(ctx, user)
+					if err != nil {
+						slog.Error("failed to update user in mattermost", slog.String("err", err.Error()))
+						continue
+					}
+					slog.Info("migrated user", slog.String("username", user.Username))
+					wg.Done()
+
+				case <-refreshTokenTicker.C:
+					slog.Info("refreshing keycloak token", slog.Int("worker", workerNumber))
+					keycloakToken, err = keycloakClient.LoginAdmin(
+						ctx,
+						deploymentConfig.ExternalAuthProviderSettings.KeycloakAdminUser,
+						deploymentConfig.ExternalAuthProviderSettings.KeycloakAdminPassword,
+						"master", // TODO: Allow specifying the master realm
+					)
+					if err != nil {
+						slog.Error("failed to refresh keycloak token", slog.String("err", err.Error()))
+						close(doneChan)
+						panic(err)
+					}
+
+				case <-doneChan:
+					return
+				}
+			}
+		}()
+	}
+
+	// File writter for users.txt
+	go func() {
+		usersTxtFile, err := os.OpenFile("users.txt", os.O_CREATE|os.O_WRONLY, 0644)
+		if err != nil {
+			slog.Error("failed to open users.txt file")
+			panic(err)
+		}
+		defer usersTxtFile.Close()
+
+		for {
+			select {
+			case email := <-usersTxtChan:
+				_, err := usersTxtFile.Write([]byte(model.UserAuthServiceSaml + ":" + email + " " + userPassword + "\n"))
+				if err != nil {
+					slog.Error("failed to write to users.txt", slog.String("err", err.Error()))
+				}
+
+			case <-doneChan:
+				return
+			}
+		}
+	}()
+
+	page := 0
+	perPage := 100
+	for {
+		slog.Info("fetching mattermost users", slog.Int("page", page), slog.Int("per_page", perPage))
+		users, _, err := mmClient.GetUsers(ctx, page, perPage, "")
+		if err != nil {
+			return fmt.Errorf("failed to get users from mattermost: %w", err)
+		}
+
+		for _, user := range users {
+			// Skip bots
+			if user.IsBot {
+				continue
+			}
+
+			// Write the user to the users.txt file
+			usersTxtChan <- user.Email
+
+			// Already migrated
+			if user.AuthService == model.UserAuthServiceSaml {
+				continue
+			}
+
+			if !dryRun {
+				wg.Add(1)
+				usersChan <- user
+			} else {
+				slog.Info("dry-run: would migrate user", slog.String("username", user.Username))
+			}
+		}
```

Comment:
This is the code which should be earlier. Because this is the one which is feeding the data to the consumer goroutines. Otherwise it's difficult to follow the flow of the code.